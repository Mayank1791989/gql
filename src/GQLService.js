/* @flow */
import * as schema from './schema';
import * as query from './query';

import GQLConfig from './config/GQLConfig';
import debug from './shared/debug';

import type {
  GQLHint,
  GQLInfo,
  DefLocation,
  Position,
} from './shared/types';

import type { GQLError } from './shared/GQLError';

type Options = $Exact<{
  cwd?: string,
  onChange?: () => void,
  onInit?: () => void,
  debug?: true,
}>;

type CommandParams = $Exact<{
  sourcePath: string,
  sourceText: string,
  position: Position,
}>;

export type { Options as GQLServiceOptions };

export class GQLService {
  _isInitialized: boolean = false;
  _schemaBuilder: schema.SchemaBuilder;
  _queryManager: ?query.QueryManager;

  _config: GQLConfig;

  constructor(options: ?Options) {
    const { onChange, onInit, debug: enableDebug, ...configOptions } = options || {};
    if (enableDebug) { (debug: any).enable(); }
    this._config = new GQLConfig(configOptions);
    this._schemaBuilder = new schema.SchemaBuilder({
      config: this._config,
      watch: true,
      onChange,
      onInit: () => {
        if (this._config.getQuery()) {
          this._queryManager = new query.QueryManager({
            config: this._config,
            getSchema: () => this._schemaBuilder.getSchema(),
            onInit: () => {
              this._isInitialized = true;
              if (onInit) { onInit(); }
            },
            onChange,
          });
        } else {
          this._isInitialized = true;
          if (onInit) { onInit(); }
        }
      },
    });
  }

  // list all file extensions defined in gqlconfig
  getFileExtensions(): Array<string> {
    return this._config.getFileExtensions();
  }

  status(): Array<GQLError> {
    if (!this._isInitialized) { return []; }
    const schemaErrors = this._schemaBuilder.getSchemaErrors();
    const queryErrors = this._queryManager ? this._queryManager.getErrors() : [];
    return schemaErrors.concat(queryErrors.filter(err => Boolean(err)));
  }

  autocomplete(params: CommandParams): Array<GQLHint> {
    debug.log('\n[autocomplete]'); debug.time('time');
    const { sourceText, sourcePath, position } = params;
    if (!this._isInitialized) { return []; }

    // codemirror instance
    let results = [];
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('autocomplete');
    if (match && match.type === 'schema') {
      results = schema.commands.getHintsAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
      );
    }

    if (match && match.type === 'query') {
      results = query.commands.getHintsAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
        match.opts,
      );
    }
    debug.timeEnd('autocomplete');
    debug.timeEnd('time');
    return results;
  }

  getDef(params: CommandParams): ?DefLocation {
    debug.log('\n[getDef]'); debug.time('time');
    const { sourceText, sourcePath, position } = params;
    if (!this._isInitialized) { return undefined; }

    let defLocation = null;
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('getDef');
    if (match && match.type === 'schema') {
      defLocation = schema.commands.getDefinitionAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
      );
    }

    if (match && match.type === 'query') {
      defLocation = query.commands.getDefinitionAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
        match.opts.parser,
      );
    }
    debug.timeEnd('getDef');
    debug.timeEnd('time');
    return defLocation;
  }

  findRefs(params: CommandParams): Array<DefLocation> {
    debug.log('\n[findRefs]'); debug.time('time');
    const { sourceText, sourcePath, position } = params;
    if (!this._isInitialized) { return []; }

    let refLocations = [];
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('findRefs');
    if (match && match.type === 'schema') {
      refLocations = schema.commands.findRefsOfTokenAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
      );
    }
    debug.timeEnd('findRefs');

    debug.timeEnd('time');
    // @TODO query not implemented
    return refLocations;
  }

  getInfo(params: CommandParams): ?GQLInfo {
    debug.log('\n[getInfo]'); debug.time('time');
    const { sourcePath, sourceText, position } = params;
    if (!this._isInitialized) { return undefined; }

    let info = null;
    debug.time('match');
    const match = this._config.match(sourcePath);
    debug.timeEnd('match');
    debug.log('FileType:', match && match.type);

    debug.time('getInfoAtToken');
    if (match && match.type === 'schema') {
      info = schema.commands.getInfoOfTokenAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
      );
    }

    if (match && match.type === 'query') {
      info = query.commands.getInfoOfTokenAtPosition(
        this._schemaBuilder.getSchema(),
        sourceText,
        position,
        match.opts,
      );
    }
    debug.timeEnd('getInfoAtToken');

    debug.timeEnd('time');
    return info;
  }
}
