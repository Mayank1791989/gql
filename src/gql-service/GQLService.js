/* @flow */
import GQLSchemaService from 'gql-schema-service';
import GQLQueryService from 'gql-query-service';
import GQLConfig from 'gql-config';
import GQLWatcher from 'gql-watcher';
import GQLBaseService from 'gql-shared/GQLBaseService';
import GQLSchema from 'gql-shared/GQLSchema';
import log from 'gql-shared/log';

const logger = log.getLogger('gql');

import {
  type GQLHint,
  type GQLInfo,
  type GQLLocation,
  type GQLPosition,
} from 'gql-shared/types';

import { type GQLError } from 'gql-shared/GQLError';
import getResolverLocation from 'gql-schema-service/commands/getResolverLocation';
import getSchemaLocation from 'gql-schema-service/commands/getSchemaLocation';

type Options = $ReadOnly<{|
  configDir?: string,

  // watch
  watchman?: boolean,
  watch?: boolean,
|}>;

type CommandParams = $ReadOnly<{|
  sourcePath: string,
  sourceText: string,
  position: GQLPosition,
|}>;

type GetResolverOrSchemaParams = $ReadOnly<{|
  fieldType: string,
  fieldName: string,
|}>;

export type { Options as GQLServiceOptions };

export function isGql(sourcePath: string) {
  return !(sourcePath.endsWith('ts') || sourcePath.endsWith('ts'));
}

export default class GQLService extends GQLBaseService {
  _schemaService: GQLSchemaService;
  _queryService: ?GQLQueryService;

  _config: GQLConfig;
  _watcher: GQLWatcher;

  constructor(_options: ?Options) {
    super();
    const options = _options || {};

    this._config = new GQLConfig({
      configDir: options.configDir || process.cwd(),
    });
    this._watcher = new GQLWatcher({
      watchman: options.watchman,
      watch: options.watch,
    });

    // setup schema service
    this._schemaService = new GQLSchemaService({
      config: this._config,
      watcher: this._watcher,
    });
    this._schemaService.onChange(this._triggerChange);
    this._schemaService.onError(this._triggerError);

    // setup query service
    if (this._config.getQueryConfig()) {
      const queryService = new GQLQueryService({
        config: this._config,
        schemaService: this._schemaService,
        watcher: this._watcher,
      });
      queryService.onChange(this._triggerChange);
      queryService.onError(this._triggerError);
      this._queryService = queryService;
    }
  }

  onLog(listener) {
    return log.onLog(listener);
  }

  async _handleStart() {
    // start schema service
    await this._schemaService.start();
    if (this._queryService) {
      await this._queryService.start();
    }
  }

  async _handleStop() {
    await this._schemaService.stop();
    if (this._queryService) {
      await this._queryService.stop();
    }
  }

  getSchema(): GQLSchema {
    return this._schemaService.getSchema();
  }

  getConfig(): GQLConfig {
    return this._config;
  }

  status(): Array<GQLError> {
    if (!this._isRunning) {
      return [];
    }
    try {
      const schemaErrors = this._schemaService.getSchemaErrors();
      const queryErrors = this._queryService
        ? this._queryService.getErrors()
        : [];
      return schemaErrors.concat(queryErrors.filter(err => Boolean(err)));
    } catch (err) {
      this._triggerError(err);
      return [];
    }
  }

  autocomplete(params: CommandParams): Array<GQLHint> {
    return this._catchThrownErrors(() => {
      logger.debug('autocomplete request');
      logger.time('autocomplete response');
      if (!this._isRunning) {
        return [];
      }

      // codemirror instance
      let results = [];
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('getHintsAtPosition');
      if (fileConfig && fileConfig.type === 'schema') {
        results = this._schemaService.getHintsAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        results = this._queryService.getHintsAtPosition({
          fileContent: params.sourceText,
          filePath: params.sourcePath,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      logger.timeEnd('getHintsAtPosition');
      logger.timeEnd('autocomplete response');
      return results;
    }, []);
  }

  getLenses(params: CommandParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getLenses request');
      logger.time('getLenses response');
      if (!this._isRunning) {
        return null;
      }

      let defLenses = null;

      if (isGql(params.sourcePath)) {
        logger.time('getFileConfig');
        const fileConfig = this._config.getFileConfig(params.sourcePath);
        logger.timeEnd('getFileConfig');
        logger.debug('FileType detected:', fileConfig && fileConfig.type);

        logger.time('getLenses');

        if (fileConfig && fileConfig.type === 'schema') {
          defLenses = await this._schemaService.getLenses({
            filePath: params.sourcePath,
          });
        }
      } else {
        logger.time('getLenses');
        defLenses = await this._schemaService.getLenses({
          filePath: params.sourcePath,
        });
      }

      logger.timeEnd('getLenses');
      logger.timeEnd('getLenses response');
      return defLenses;
    }, null);
  }

  getResolverLocation(params: GetResolverOrSchemaParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getResolverLocation request');
      logger.time('getResolverLocation response');
      if (!this._isRunning) {
        return null;
      }

      logger.time('getResolverLocation');

      const schema = this._schemaService.getSchema();
      const resolverLocation = await getResolverLocation(
        params.fieldType,
        params.fieldName,
        schema,
        this._config.getResolverConfig().files,
        this._config.getResolverConfig().baseDir,
      );

      logger.timeEnd('getResolverLocation');
      logger.timeEnd('getResolverLocation response');
      return resolverLocation;
    }, null);
  }

  getSchemaLocation(params: GetResolverOrSchemaParams): ?GQLLocation {
    return this._catchThrownErrors(() => {
      logger.debug('getSchemaLocation request');
      logger.time('getSchemaLocation response');
      if (!this._isRunning) {
        return null;
      }

      logger.time('getSchemaLocation');

      const schema = this._schemaService.getSchema();
      const schemaLocation = getSchemaLocation(
        params.fieldType,
        params.fieldName,
        schema,
      );

      logger.timeEnd('getSchemaLocation');
      logger.timeEnd('getSchemaLocation response');
      return schemaLocation;
    }, null);
  }

  getDef(params: CommandParams): ?GQLLocation {
    return this._catchThrownErrors(async () => {
      logger.debug('getDef request');
      logger.time('getDef response');
      if (!this._isRunning) {
        return null;
      }

      let defLocation = null;

      if (isGql(params.sourcePath)) {
        logger.time('getFileConfig');
        const fileConfig = this._config.getFileConfig(params.sourcePath);
        logger.timeEnd('getFileConfig');
        logger.debug('FileType detected:', fileConfig && fileConfig.type);

        logger.time('getDef');
        if (fileConfig && fileConfig.type === 'schema') {
          defLocation = await this._schemaService.getDefinitionAtPosition({
            fileContent: params.sourceText,
            filePath: params.sourcePath,
            fileOptions: fileConfig.opts,
            position: params.position,
            resolversGlob: this._config.getResolverConfig().files,
            resolversBaseDir: this._config.getResolverConfig().baseDir,
          });
        }

        if (this._queryService && fileConfig && fileConfig.type === 'query') {
          defLocation = await this._queryService.getDefinitionAtPosition({
            fileContent: params.sourceText,
            filePath: params.sourcePath,
            fileOptions: fileConfig.opts,
            position: params.position,
          });
        }
      } else {
        logger.time('getDef');
        defLocation = await this._schemaService.getDefinitionAtPosition({
          fileContent: params.sourceText,
          filePath: params.sourcePath,
          position: params.position,
        });
      }
      logger.timeEnd('getDef');
      logger.timeEnd('getDef response');
      return defLocation;
    }, null);
  }

  findRefs(params: CommandParams): Array<GQLLocation> {
    return this._catchThrownErrors(() => {
      logger.debug('findRefs request');
      logger.time('findRefs response');
      if (!this._isRunning) {
        return [];
      }

      let refLocations = [];
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('findRefs');
      if (fileConfig && fileConfig.type === 'schema') {
        refLocations = this._schemaService.findRefsOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }
      // if (this._queryService && fileConfig && fileConfig.type === 'query') {
      //   refLocations = this._queryService.findRefsOfTokenAtPosition({
      //     fileContent: params.sourceText,
      //     filePath: params.sourcePath,
      //     fileOptions: fileConfig.opts,
      //     position: params.position,
      //   });
      // }
      logger.timeEnd('findRefs');
      logger.timeEnd('findRefs response');
      // @TODO query not implemented
      return refLocations;
    }, []);
  }

  getInfo(params: CommandParams): ?GQLInfo {
    return this._catchThrownErrors(() => {
      logger.debug('getInfo request');
      logger.time('getInfo response');
      if (!this._isRunning) {
        return null;
      }

      let info = null;
      logger.time('getFileConfig');
      const fileConfig = this._config.getFileConfig(params.sourcePath);
      logger.timeEnd('getFileConfig');
      logger.debug('FileType detected:', fileConfig && fileConfig.type);

      logger.time('getInfoAtToken');
      if (fileConfig && fileConfig.type === 'schema') {
        info = this._schemaService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          position: params.position,
        });
      }

      if (this._queryService && fileConfig && fileConfig.type === 'query') {
        info = this._queryService.getInfoOfTokenAtPosition({
          fileContent: params.sourceText,
          fileOptions: fileConfig.opts,
          filePath: params.sourcePath,
          position: params.position,
        });
      }
      logger.timeEnd('getInfoAtToken');
      logger.timeEnd('getInfo response');
      return info;
    }, null);
  }
}
