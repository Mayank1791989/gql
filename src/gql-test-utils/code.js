/* @flow */
import { dedent } from 'dentist';
import splitLines from 'gql-shared/splitLines';

export default function code(text: string) {
  const sourceText = dedent(text);
  const lines = splitLines(sourceText);
  let position = null;
  lines.forEach((line, index) => {
    const match = line.match(/--\^/);
    if (match) {
      position = {
        // $FlowDisableNextLine
        column: match.index + 3,
        line: index,
      };
    }
  });

  if (!position) {
    throw new Error('Missing --^ in source');
  }

  return {
    sourceText,
    position,
  };
}
