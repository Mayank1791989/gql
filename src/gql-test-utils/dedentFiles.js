/* @flow */
import { dedent } from 'dentist';

type FilesMap = { [path: string]: string };

export function dedentFiles(files?: FilesMap): FilesMap {
  if (!files) {
    return {};
  }
  return Object.keys(files).reduce((acc, path) => {
    acc[path] = dedent(files[path]);
    return acc;
  }, {});
}
