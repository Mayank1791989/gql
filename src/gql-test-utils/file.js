/* @flow */
import os from 'os';
import objectHash from 'object-hash';
import { writeFileSync } from 'gql-shared/fs-utils';
import { execSync } from 'child_process';
import path from 'path';

export const TEST_TEMP_DIR = path.join(os.tmpdir(), 'gql-test');

export function createTempFiles(files: { [file: string]: string }): string {
  // deterministic dirName [as this path used in test cases output (see generated snapshot)]
  const dir = path.join(TEST_TEMP_DIR, objectHash(files));
  Object.keys(files).forEach(file => {
    const content = files[file];
    writeFileSync(path.join(dir, file), content);
  });
  return dir;
}

export function removeTestTempDir() {
  execSync(`rm -rf ${TEST_TEMP_DIR}`);
}
