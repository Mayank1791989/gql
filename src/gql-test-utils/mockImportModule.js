/* @flow */
export default function mockImportModule(
  moduleName: string,
  mockFactory: () => $FixMe,
) {
  jest.doMock('import-from', () => {
    return function importFrom(dir, moduleID) {
      if (moduleID === moduleName) {
        return mockFactory();
      }
      const importFromActual = jest.requireActual('import-from');
      return importFromActual(dir, moduleID);
    };
  });
}
