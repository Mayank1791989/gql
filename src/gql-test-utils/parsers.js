/* @flow */
import SchemaParser from 'gql-schema-parser-default';
import EmbeddedQueryParser from 'gql-query-parser-embedded-queries';
import QueryParser from 'gql-query-parser-default';
import QueryPresetRelay from 'gql-query-preset-relay';

export function relayQLParser() {
  return new EmbeddedQueryParser({
    start: 'Relay\\.QL`',
    end: '`',
    allowFragmentWithoutName: true,
    allowFragmentInterpolation: true,
  });
}

export function apolloParser() {
  return new EmbeddedQueryParser({
    start: 'gql`',
    end: '`',
    allowDocumentInterpolation: true,
  });
}

export const relayExtendSchema = QueryPresetRelay().extendSchema;

export function queryParser() {
  return new QueryParser();
}

export function customParser() {
  return new EmbeddedQueryParser({
    start: '"""',
    end: '"""',
  });
}

export function embeddedQueryParser({
  start,
  end,
}: {
  start: string,
  end: string,
}) {
  return new EmbeddedQueryParser({
    start,
    end,
  });
}

export function schemaParser() {
  return new SchemaParser();
}
