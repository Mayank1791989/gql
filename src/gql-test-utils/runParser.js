/* @flow */
import { type IParser } from 'gql-shared/types';
import MultilineCharacterStream from 'gql-shared/MultilineCharacterStream';

export default function runParser({
  sourceText,
  parser,
}: {
  sourceText: string,
  parser: IParser,
}) {
  const state = parser.startState();
  const stream = new MultilineCharacterStream(sourceText);
  const result = [];
  while (!stream.eof()) {
    const style = parser.token(stream, state);
    result.push({
      start: stream.getStartOfToken(),
      end: stream.getCurrentPosition(),
      text: stream.current(),
      style,
    });
  }

  return result;
}
