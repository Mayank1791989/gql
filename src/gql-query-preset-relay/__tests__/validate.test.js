/* @flow */
import { createTempFiles } from 'gql-test-utils/file';
import { dedent } from 'dentist';
import GQLService from 'gql-service';

async function validateSource(source: string): Promise<any> {
  const gql = new GQLService({
    configDir: createTempFiles({
      '.gqlconfig': `
        {
          schema: {
            files: 'schema/*.gql',
            validate: {
              config: {
                NoUnusedTypeDefinition: 'off'
              }
            }
          },

          query: {
            files: [
              {
                match: 'query/*.js',
                presets: ['relay']
              }
            ]
          }
        }
      `,

      'schema/schema.gql': `
        type Query {
          viewer: Viewer!
          mutation: Mutation!
        }

        type Viewer {
          me: User!
        }

        type User {
          id: ID!
          name: String
          image(size: Int!): String
        }

        type Mutation {
          UserCreate(input: UserCreateInput): UserCreatePayload
        }

        input UserCreateInput {
          id: ID!
          name: String!
        }

        type UserCreatePayload {
          userID: ID!
          viewer: Viewer!
        }
      `,

      'query/test.js': dedent(source),
    }),

    watch: false,
  });

  gql.onError(err => {
    throw err;
  });

  await gql.start();

  return gql.status();
}

test('report error for missing subselection in fragments', async () => {
  const errors = await validateSource(`
    Relay.QL\`
      fragment on Viewer {
        me
      }
    \`
  `);
  expect(errors).toMatchSnapshot();
});

test('allow field without subselection in fragments if @relay(pattern: true) is present', async () => {
  const errors = await validateSource(`
    Relay.QL\`
      fragment on UserCreatePayload @relay(pattern: true) {
        viewer
      }
    \`
  `);
  expect(errors).toMatchSnapshot();
});

test('allow query without subselection', async () => {
  const errors = await validateSource(`
    Relay.QL\`
      query test { viewer }
    \`
  `);
  expect(errors).toMatchSnapshot();
});

test('allow mutation without subselection', async () => {
  const errors = await validateSource(`
    Relay.QL\`
      mutation tester { UserCreate }
    \`
  `);
  expect(errors).toMatchSnapshot();
});

describe('Rule: ArgumentsOfCorrectType', () => {
  it('report error for wrong type', async () => {
    const errors = await validateSource(`
      Relay.QL\`
        fragment on User {
          image(size: "some_string_value")
        }
      \`
    `);
    expect(errors).toMatchSnapshot();
  });
});

describe('Rule: ProvidedNonNullArguments', () => {
  it('report error if field arguments missing', async () => {
    const errors = await validateSource(`
      Relay.QL\`
        fragment on User {
          image
        }
      \`
    `);
    expect(errors).toMatchSnapshot();
  });

  it('allow mutation field without arguments', async () => {
    const errors = await validateSource(`
      Relay.QL\`
        mutation test { UserCreate }
      \`
    `);
    expect(errors).toMatchSnapshot();
  });
});

test('should not report errors for valid query', async () => {
  const errors = await validateSource(`
    Relay.QL\`
      fragment on User {
        name
        id
      }
    \`
  `);
  expect(errors).toMatchSnapshot();
});

test('report if operation name missing', async () => {
  const errors = await validateSource(`
    Relay.QL\`
      query {
        viewer
      }
    \`
  `);
  expect(errors).toMatchSnapshot();
});

describe('[Rule] Exactly One Definition', () => {
  test('valid', async () => {
    const errors = await validateSource(`
      Relay.QL\`
        fragment Test on User {
          name
          id
        }
      \`
    `);
    expect(errors).toEqual([]);
  });

  test('invalid', async () => {
    const errors = await validateSource(`
      import Relay from 'react-relay';

      const a = Relay.QL\`
        fragment Test on User {
          name
          id
        }

        fragment Test2 on User {
          name
        }
      \`
    `);
    expect(errors).toMatchSnapshot();
  });
});

test('[Rule] ExecutableDefinitions', async () => {
  const errors = await validateSource(`
    const a = Relay.QL\`
      type Tester {
        name: String
      }
    \`
  `);
  expect(errors).toMatchSnapshot();
});
