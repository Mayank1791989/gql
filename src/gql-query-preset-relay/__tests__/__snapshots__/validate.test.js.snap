// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Rule] Exactly One Definition invalid 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 4,
        "path": "$TEST_TEMP_DIR/2c178f90d07db92c2eaa7e2dd8610735b8491f87/query/test.js",
      },
      Object {
        "column": 3,
        "line": 9,
        "path": "$TEST_TEMP_DIR/2c178f90d07db92c2eaa7e2dd8610735b8491f87/query/test.js",
      },
    ],
    "message": "Expected exactly one definition. (ExactlyOneDefinition)",
    "severity": "error",
  },
]
`;

exports[`[Rule] ExecutableDefinitions 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 2,
        "path": "$TEST_TEMP_DIR/5c79f83b15a4e324396e3fc7e002a18ae31f5bc2/query/test.js",
      },
    ],
    "message": "Definition is not executable (Only fragment, mutation, query or subscription is allowed). (ExecutableDefinitions)",
    "severity": "error",
  },
]
`;

exports[`Rule: ArgumentsOfCorrectType report error for wrong type 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 17,
        "line": 3,
        "path": "$TEST_TEMP_DIR/f1e9f40e9c127eb7820c5fc50c215350e3dbbfab/query/test.js",
      },
    ],
    "message": "Expected type Int!, found \\"some_string_value\\". (ValuesOfCorrectType)",
    "severity": "error",
  },
]
`;

exports[`Rule: ProvidedNonNullArguments allow mutation field without arguments 1`] = `Array []`;

exports[`Rule: ProvidedNonNullArguments report error if field arguments missing 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 5,
        "line": 3,
        "path": "$TEST_TEMP_DIR/2475cd5106a22dd238e79ca93bdb0138f65761fd/query/test.js",
      },
    ],
    "message": "Field \\"image\\" argument \\"size\\" of type \\"Int!\\" is required but not provided. (ProvidedNonNullArguments)",
    "severity": "error",
  },
]
`;

exports[`allow field without subselection in fragments if @relay(pattern: true) is present 1`] = `Array []`;

exports[`allow mutation without subselection 1`] = `Array []`;

exports[`allow query without subselection 1`] = `Array []`;

exports[`report error for missing subselection in fragments 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 5,
        "line": 3,
        "path": "$TEST_TEMP_DIR/ac23a5be7f29c150dfb1776aac3f026173924946/query/test.js",
      },
    ],
    "message": "Field \\"me\\" of type \\"User!\\" must have a selection of subfields. Did you mean \\"me { ... }\\"? (ScalarLeafs)",
    "severity": "error",
  },
]
`;

exports[`report if operation name missing 1`] = `
Array [
  Object {
    "locations": Array [
      Object {
        "column": 3,
        "line": 2,
        "path": "$TEST_TEMP_DIR/289c472df0d60fed029911c980ee9299c8783736/query/test.js",
      },
    ],
    "message": "query name is missing. (RequiredOperationName)",
    "severity": "warn",
  },
]
`;

exports[`should not report errors for valid query 1`] = `Array []`;
